internal/directives/json_updater.go

```diff
@@ -0,0 +1,145 @@
+package directives
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+	"github.com/tidwall/sjson"
+	"github.com/xeipuuv/gojsonschema"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+)
+
+func init() {
+	builtins.RegisterPromotionStepRunner(newJSONUpdater(), nil)
+}
+
+// jsonUpdater is an implementation of the PromotionStepRunner interface that
+// updates the values of specified keys in a JSON file.
+type jsonUpdater struct {
+	schemaLoader gojsonschema.JSONLoader
+}
+
+// newJSONUpdater returns an implementation of the PromotionStepRunner interface
+// that updates the values of specified keys in a JSON file.
+func newJSONUpdater() PromotionStepRunner {
+	r := &jsonUpdater{}
+	r.schemaLoader = getConfigSchemaLoader(r.Name())
+	return r
+}
+
+// Name implements the PromotionStepRunner interface.
+func (j *jsonUpdater) Name() string {
+	return "json-update"
+}
+
+// RunPromotionStep implements the PromotionStepRunner interface.
+func (j *jsonUpdater) RunPromotionStep(
+	ctx context.Context,
+	stepCtx *PromotionStepContext,
+) (PromotionStepResult, error) {
+	failure := PromotionStepResult{Status: kargoapi.PromotionPhaseErrored}
+
+	if err := j.validate(stepCtx.Config); err != nil {
+		return failure, err
+	}
+
+	cfg, err := ConfigToStruct[JSONUpdateConfig](stepCtx.Config)
+	if err != nil {
+		return failure, fmt.Errorf("could not convert config into %s config: %w", j.Name(), err)
+	}
+
+	return j.runPromotionStep(ctx, stepCtx, cfg)
+}
+
+// validate validates jsonUpdater configuration against a JSON schema.
+func (j *jsonUpdater) validate(cfg Config) error {
+	return validate(j.schemaLoader, gojsonschema.NewGoLoader(cfg), j.Name())
+}
+
+func (j *jsonUpdater) runPromotionStep(
+	_ context.Context,
+	stepCtx *PromotionStepContext,
+	cfg JSONUpdateConfig,
+) (PromotionStepResult, error) {
+	result := PromotionStepResult{Status: kargoapi.PromotionPhaseSucceeded}
+
+	if len(cfg.Updates) > 0 {
+		if err := j.updateFile(stepCtx.WorkDir, cfg.Path, cfg.Updates); err != nil {
+			return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+				fmt.Errorf("JSON file update failed: %w", err)
+		}
+
+		if commitMsg := j.generateCommitMessage(cfg.Path, cfg.Updates); commitMsg != "" {
+			result.Output = map[string]any{
+				"commitMessage": commitMsg,
+			}
+		}
+	}
```

Comment:
Nit: I'd remove the conditional. We validated up front that there was at least one update to perform and even if somehow the updates passed into here were nil or empty, it's still perfectly safe to iterate over them.